### PR TITLE
Use Store Actions - Initial Version

### PR DESCRIPTION
**Pull request by JamesSCBD**
1. Written store actions (contact - initial, organization - partial)
2. used store actions in reference files

### CODE REVIEW

File: components/list/OrganizationsList.vue [hunk 1]
```diff
@@ -193,7 +194,7 @@ async function rowDoubleClicked(item) {
 async function getSelectedContactsResult(task) {
   const query = this.baseQuery || {};
   if (!_.isEmpty(query)) {
-    const result               = await this.$kronosApi.queryOrganizations({ ...query, limit: 5 });
+    const result               = await this.queryOrganizations({ ...query, limit: 5 });
```
Comment (sbilodeau) [inline components/list/OrganizationsList.vue:197]: I think the name of the function above it wrong. shoulf be `getSelectedOrganizationsResult` instead of `getSelectedContactsResult`

File: components/controls/selectors/OrganizationContactSelector.vue [hunk 2]
```diff
@@ -82,14 +83,17 @@ export default {
...
+    }),
   },
 };
 
 function fullName({ firstName, lastName, designation }) {
   return _d.remove(`${firstName} ${lastName} ${designation && `(${designation.trim()})`}`);
 }
 async function load() {
-  const contacts      = this.disabled ? [] : await this.$kronosApi.getOrganizationContacts(this.organizationId);
+  const contacts      = this.disabled ? [] : await this.getOrganizationContacts(this.organizationId);
```
Comment (sbilodeau) [inline components/controls/selectors/OrganizationContactSelector.vue:96]: shoudl pass useCache=true
```suggestion
  const contacts      = this.disabled ? [] : await this.getOrganizationContacts(this.organizationId, true);
```

File: pages/organizations/_organizationId/contacts/_contactId/index.vue [hunk 3]
```diff
@@ -447,15 +453,15 @@ async function load() {
...
+      : await this.getContact(contactId);
 
     // redirect to correct url if organization defers
     if (contact.organizationId !== organizationId) {
       this.refreshUrl(contact);
     }
 
     // fetch organization details
-    const organization = await this.$kronosApi.getOrganization(organizationId);
+    const organization = await this.getOrganization(organizationId);
```
Comment (sbilodeau) [inline pages/organizations/_organizationId/contacts/_contactId/index.vue:464]: can read from cache
```suggestion
    const organization = await this.getOrganization(organizationId, true);
```

File: store/contacts.js [hunk 4]
```diff
@@ -72,6 +80,42 @@ const $actions = {
     commit(CLEAR);
   },
 
+  async queryContacts({ commit }, query) {
+    const contacts = await this.$kronosApi.queryContacts(query);
+    console.log(commit); // NEED TO REMOVE
+    return _.cloneDeep(contacts);
```
Comment (sbilodeau) [inline store/contacts.js:86]: no need to clone I think. contact is not from cache object

File: store/contacts.js [hunk 5]
```diff
@@ -72,6 +80,42 @@ const $actions = {
...
+  },
+
+  async getContactById({ commit }, id) {
+    const contact = await this.$kronosApi.getContact(id);
+    console.log(commit); // NEED TO REMOVE
+    return _.cloneDeep(contact);
```
Comment (sbilodeau) [inline store/contacts.js:92]: no need to clone I think

File: store/contacts.js [hunk 6]
```diff
@@ -72,6 +80,42 @@ const $actions = {
...
+  },
+
+  async createContact({ commit }, { organizationId, contact }) {
+    const result = await this.$kronosApi.createContact(organizationId, contact);
+    console.log(commit);  // NEED TO REMOVE
+    // TODO
+    // commit('UPDATE_ORGANIZATION_CONTACTS_CACHE', result.organizationId, result,
```
Comment (sbilodeau) [inline store/contacts.js:99]: drop organization-contacts cache

File: store/contacts.js [hunk 7]
```diff
@@ -72,6 +80,42 @@ const $actions = {
...
+    // { root: true });
+
+    return _.cloneDeep(result);
```
Comment (sbilodeau) [inline store/contacts.js:102]: no need to clone

File: store/contacts.js [hunk 8]
```diff
@@ -72,6 +80,42 @@ const $actions = {
...
+
+    console.log(commit); // NEED TO REMOVE
+    if (result) {
+      commit(UPDATE_TO_SELECTION, result);
+      // TODO
+      // commit('organizations/UPDATE_ORGANIZATION_CONTACTS_CACHE', result.organizationId,
+      // result, { root: true });
+    }
+
+    return _.cloneDeep(result);
```
Comment (sbilodeau) [inline store/contacts.js:116]: non need to clone

File: store/organizations.js [hunk 9]
```diff
@@ -41,14 +62,28 @@ const $mutations = {
...
+    // Delete oldest one when cache size > 100
+    // TODO
+    delete state.organizationContacts[organizationId];
+
+    if (contacts && contacts.length > 0) state.organizationContacts[organizationId] = contacts;
+    if (Object.keys(state.organizationContacts).length > 100) {
+      delete state.organizationContacts[Object.keys(state.organizationContacts)[0]];
+    }
+  },
+
```
Comment (sbilodeau) [inline store/organizations.js:81]: us array as you suggested
make `100` a constante at the top of the file.

File: store/organizations.js [hunk 10]
```diff
@@ -109,15 +156,55 @@ const $actions = {
...
+    let organization = {};
+
+    if (getters.isOrganizationCached(id)) {
+      organization = getters.getOrganizationById(id);
+    } else {
+      organization = await this.$kronosApi.getOrganization(id);
+      if (organization) commit(UPDATE_CACHE, [ organization ]);
+    }
+
+    return _.cloneDeep(organization);
```
Comment (sbilodeau) [inline store/organizations.js:172]: missing allowReadFromCache=false
```suggestion
  async getOrganizationById({ commit, getters }, id, allowReadFromCache=false) {
    let organization = null

    if (allowReadFromCache && getters.isOrganizationCached(id)) {
      organization = getters.getOrganizationById(id);
    } else {
      organization = await this.$kronosApi.getOrganization(id);
      if (organization) commit(UPDATE_CACHE, [ organization ]);
    }

    return _.cloneDeep(organization);
```

File: store/organizations.js [hunk 11]
```diff
@@ -109,15 +156,55 @@ const $actions = {
 
     if (organizations) commit(UPDATE_CACHE, organizations.records);
 
-    return organizations;
+    return _.cloneDeep(organizations);
+  },
+
+  async getOrganizationById({ commit, getters }, id) {
+    let organization = {};
```
Comment (sbilodeau) [inline store/organizations.js:163]: default value must be `null` not an `empty object`;

File: store/organizations.js [hunk 12]
```diff
@@ -109,15 +156,55 @@ const $actions = {
...
 
-    return organization;
+    return _.cloneDeep(result);
+  },
+
+  async updateOrganization({ commit }, organization) {
+    const result = await this.$kronosApi.updateOrganization(organization);
+    if (result) {
+      commit(UPDATE_TO_SELECTION, result);
+      commit(UPDATE_CACHE, [ result ]);
```
Comment (sbilodeau) [inline store/organizations.js:187]: I think `UPDATE_CACHE` should also do the `UPDATE_TO_SELECTION`

**Review by sbilodeau — CHANGES_REQUESTED** [review]
Take adavanage of the `allowRreadFromCache` in code when relevant

File: components/list/OrganizationsList.vue (hunk 1 above)
Comment (JamesSCBD) [reply components/list/OrganizationsList.vue:197]: DONE! changes done as suggested.

File: components/controls/selectors/OrganizationContactSelector.vue (hunk 2 above)
Comment (JamesSCBD) [reply components/controls/selectors/OrganizationContactSelector.vue]: DONE! changes done as suggested. passed parameter in one object form. as store actions take parameter in single object.

File: pages/organizations/_organizationId/contacts/_contactId/index.vue (hunk 3 above)
Comment (JamesSCBD) [reply pages/organizations/_organizationId/contacts/_contactId/index.vue]: DONE! changes done as suggested.

File: store/contacts.js (hunk 4 above)
Comment (JamesSCBD) [reply store/contacts.js]: DONE! changes done as suggested.

File: store/contacts.js (hunk 5 above)
Comment (JamesSCBD) [reply store/contacts.js]: DONE! changes done as suggested.

File: store/contacts.js (hunk 7 above)
Comment (JamesSCBD) [reply store/contacts.js]: DONE! changes done as suggested.

File: store/contacts.js (hunk 8 above)
Comment (JamesSCBD) [reply store/contacts.js]: DONE! changes done as suggested.

File: store/organizations.js (hunk 9 above)
Comment (JamesSCBD) [reply store/organizations.js]: DONE! changes done as suggested.

File: store/organizations.js (hunk 10 above)
Comment (JamesSCBD) [reply store/organizations.js]: DONE! changes done as suggested.

File: store/organizations.js (hunk 11 above)
Comment (JamesSCBD) [reply store/organizations.js]: DONE! changes done as suggested.

File: store/contacts.js (hunk 6 above)
Comment (JamesSCBD) [reply store/contacts.js]: Hi @sbilodeau, currently I don't think that we need to drop organization-contacts at the time of adding/updating contact of organization. So, I have updated organization-cache when any contacts is added/updated.

File: store/organizations.js (hunk 12 above)
Comment (JamesSCBD) [reply store/organizations.js]: DONE! changes done as suggested. UPDATE_CACHE will update all organization related cache and REMOVE_CACHE will remove all organization related cache

File: store/contacts.js [hunk 13]
```diff
@@ -7,18 +7,30 @@ const $state = () => ({
...
 const REMOVE_FROM_SELECTION = 'REMOVE_FROM_SELECTION';
 const CLEAR                 = 'CLEAR';
 const CLEAR_SELECTION       = 'CLEAR_SELECTION';
+const NO_ACTION             = 'NO_ACTION';
 
 const $mutations = {
 
+  // call this mutation if don't want to commit anything
+  [NO_ACTION]() {},
+
```
Comment (sbilodeau) [inline store/contacts.js:20]: why? 
if you have nothing to commit than you don't commit anything 

Delete this mutation please.

File: store/contacts.js [hunk 14]
```diff
@@ -72,6 +84,43 @@ const $actions = {
     commit(CLEAR);
   },
 
+  async queryContacts({ commit }, query) {
+    const contacts = await this.$kronosApi.queryContacts(query);
+    commit(NO_ACTION);
```
Comment (sbilodeau) [inline store/contacts.js:89]: If you have nothing to commit then you don't commit anything.

Probably you should update cache with returned objects instead. no?,

File: store/contacts.js [hunk 15]
```diff
@@ -72,6 +84,43 @@ const $actions = {
...
+    return contacts;
+  },
+
+  async getContact({ commit }, id) {
+    const contact = await this.$kronosApi.getContact(id);
+    commit(NO_ACTION);
```
Comment (sbilodeau) [inline store/contacts.js:95]: You should update cache / selections with the new fresh returned object.

don't commit nothing. this trigger useless app refresh

File: store/organizations.js [hunk 16]
```diff
@@ -23,6 +28,29 @@ const $mutations = {
 
   [UPDATE_CACHE](state, organizations = []) {
     state.organizationCache = _.unionBy(organizations, state.organizationCache, (o) => o.organizationId);
+    _.forEach(organizations, (organization) => {
```
Comment (sbilodeau) [inline store/organizations.js:31]: you can also use native forEach of javascript array
```suggestion
    organizations.forEach((organization) => {
```

File: store/organizations.js [hunk 17]
```diff
@@ -23,6 +28,29 @@ const $mutations = {
...
+      if (organization && organization.organizationId) {
+        const index = state.selectedOrganizations.findIndex((o) => o.organizationId === organization.organizationId);
+        if (index >= 0) state.selectedOrganizations.splice(index, 1, organization);
+      }
```
Comment (sbilodeau) [inline store/organizations.js:35]: remove this if condition there. you cannot pass and array without organization inside. this check prevent good code elsewhere

add separate validation check at line 30 that test that the passed array !=null * evry value are valid. if not throw error.

eg: line 30
```javascript
// 1st line of the Methods (line 30)
const isValid = organizations && organizations.every((o)=> o && o.orgnizationId);
if(!isValid) throw Error("organizations list is invalid");
```

then do operations

 ```suggestion
      const index = state.selectedOrganizations.findIndex((o) => o.organizationId === organization.organizationId);
      if (index >= 0) state.selectedOrganizations.splice(index, 1, organization);
```

File: store/organizations.js [hunk 18]
```diff
@@ -41,14 +69,48 @@ const $mutations = {
...
-  [CLEAR](state) {
-    state.organizationCache     = [];
-    state.organizationTypes     = [];
+  [CLEAR_SELECTION](state) {
     state.selectedOrganizations = [];
   },
 
-  [CLEAR_SELECTION](state) {
+  // pass contact if contact is added or updated
+  [UPDATE_ORGANIZATION_CONTACTS_CACHE](state, { organizationId, contacts, contact }) {
```
Comment (sbilodeau) [inline store/organizations.js:77]: The current implementation is way too complex for nothing.

why having two non exclusive parameters???  I did not read it all I was lost after two lines

switch back to initial idea: Drop cached value

And only allow params: `organizationId` (madatory)  && `contacts` (null/array);

```javascript 
organizationContactCache = organizationContactCache.filter(o=>o.organizationId!=organizationId) //Systematically drop org

if(contacts) { //
  organizationContactCache.push({organizationId, contacts}) // add at the end
}

while (organizationContactCache.length> MAX_ORGANIZATION_CONTACT_CACHE) // keep only maz size latest orgs
  organizationContactCache.shift() // remove 1st element in the array
```

 

if(

File: store/organizations.js [hunk 19]
```diff
@@ -109,15 +187,52 @@ const $actions = {
 
     if (organizations) commit(UPDATE_CACHE, organizations.records);
 
-    return organizations;
+    return _.cloneDeep(organizations);
+  },
+
+  async getOrganization({ commit, getters }, { id, allowReadFromCache }) {
```
Comment (sbilodeau) [inline store/organizations.js:193]: Change method signature for the one specified int the trello card.
Using two parameters. not an object 
```suggestion
  async getOrganization({ commit, getters },  id, allowReadFromCache=false) {
```

File: store/organizations.js [hunk 20]
```diff
@@ -109,15 +187,52 @@ const $actions = {
...
 
-    if (organization) commit(UPDATE_CACHE, [ organization ]);
+  async deleteOrganization({ commit }, organizationId) {
+    await this.$kronosApi.deleteOrganization(organizationId);
 
-    return organization;
+    commit(REMOVE_CACHE, organizationId);
+  },
+
+  async getOrganizationContacts({ commit, getters }, { id, allowReadFromCache }) {
```
Comment (sbilodeau) [inline store/organizations.js:228]: Change method signature for the one specified int the trello card.
Using two parameters and default value. not an object
```suggestion
  async getOrganizationContacts({ commit, getters }, id, allowReadFromCache=false) {
```

File: store/organizations.js [hunk 21]
```diff
@@ -41,14 +69,48 @@ const $mutations = {
...
+    if (!oldCache || oldCache.length <= 0) return;
+
+    const oldCacheOrganization = oldCache.find((oc) => oc.organizationId === organizationId);
+    if (!oldCacheOrganization) return;
+
+    const oldCacheContacts     = oldCacheOrganization.contacts;
+    const contacts             = oldCacheContacts.filter((c) => c.contactId !== contactId);
+    const updatedEntry         = { organizationId, contacts };
+    state.organizationContacts = _([ updatedEntry ]).unionBy(oldCache, (o) => o.organizationId).value();
+  },
```
Comment (sbilodeau) [inline store/organizations.js:107]: Drop this mutation and use UPDATE_ORGANIZATION_CONTACTS_CACHE({ organizationId, contacts=null})

File: store/organizations.js [hunk 22]
```diff
@@ -109,15 +187,52 @@ const $actions = {
...
-    return organization;
+    commit(REMOVE_CACHE, organizationId);
+  },
+
+  async getOrganizationContacts({ commit, getters }, { id, allowReadFromCache }) {
+    const contacts = allowReadFromCache && getters.isOrganizationContactsCached(id)
+      ? getters.getOrganizationContacts(id)
+      : await this.$kronosApi.getOrganizationContacts(id);
+
+    commit(UPDATE_ORGANIZATION_CONTACTS_CACHE, { organizationId: id, contacts });
```
Comment (sbilodeau) [inline store/organizations.js:233]: Only update cache if read from api
```suggestion
    let contacts = null
    
    if(allowReadFromCache && getters.isOrganizationContactsCached(id) {
      contacts = getters.getOrganizationContacts(id)
    } else {
      contacts = await this.$kronosApi.getOrganizationContacts(id);
      commit(UPDATE_ORGANIZATION_CONTACTS_CACHE, { organizationId: id, contacts });
    }
```

**Review by sbilodeau — CHANGES_REQUESTED** [review]
Remove Organization cahce update to simple "drop" as proposed in the card.
The implementation is too compleax an I saw many flaws already and many other possible one. Also code was too complexe

Follow requirement in method signature from trello card:

required: 
```javascript
getOrganizationContacts (id, allowCache=false) ///with default value param for allowCache
```

but you implemented:
```javascript
 getOrganizationContacts ({id, allowCache}) //passing object instead of params, require to pass object everywhere insteadOf normal call with 'optional' paramter
```

File: store/contacts.js (hunk 13 above)
Comment (JamesSCBD) [reply store/contacts.js:20]: this is because I don't know how to disable eslint for paticular line. I was facing error because we don't want to use commit in method and 1st argument is required for context so....

Comment (JamesSCBD) [reply store/contacts.js:20]: Deleted unwanted mutation (NO_ACTION)

File: store/organizations.js (hunk 20 above)
Comment (JamesSCBD) [reply store/organizations.js:228]: Hi @sbilodeau, 

Actually vuex action will only accept two parameters... One is context and second is payload... 

if we pass more parameters we will get undefined value of those parameters... 

So, I have passed it as a object.

File: store/contacts.js (hunk 14 above)
Comment (JamesSCBD) [reply store/contacts.js]: Removed NO_ACTION mutation

File: store/contacts.js (hunk 15 above)
Comment (JamesSCBD) [reply store/contacts.js]: Removed NO_ACTION mutation, and commited UPDATE_CACHE.

File: store/organizations.js (hunk 16 above)
Comment (JamesSCBD) [reply store/organizations.js]: Used native foreach

File: store/organizations.js (hunk 17 above)
Comment (JamesSCBD) [reply store/organizations.js]: Hi @sbilodeau, Added separate validation and removed loop. 

Had done this because we don't require to loop through organizations array twice (One for validate, one for actual operation).

File: store/organizations.js (hunk 18 above)
Comment (JamesSCBD) [reply store/organizations.js]: Dropped organization cache at time of add/update/delete contact

File: store/organizations.js (hunk 21 above)
Comment (JamesSCBD) [reply store/organizations.js]: Dropped mutation and used suggested mutation

File: store/organizations.js (hunk 22 above)
Comment (JamesSCBD) [reply store/organizations.js]: DONE! changes done as suggested

File: store/organizations.js (hunk 20 above)
Comment (sbilodeau) [reply store/organizations.js]: ok
